roles/autoscale/README.md

```diff
@@ -0,0 +1,109 @@
+
+## stackhpc.openhpc role variables
+This role modifies what the [openhpc_slurm_partitions variable](https://github.com/stackhpc/ansible-role-openhpc#slurmconf) in the `stackhpc.openhpc` role accepts. Partition/group definitions may additionally include:
+- `cloud_nodes`: Optional. Slurm hostlist expression (e.g. `'small-[8,10-16]'`) defining names of nodes to be defined in a ["CLOUD" state](https://slurm.schedmd.com/slurm.conf.html#OPT_CLOUD), i.e. not operational when the Slurm control daemon starts.
+- `cloud_instances`: Required if `cloud_nodes` is defined. A mapping with keys `flavor`, `image`, `keypair` and `network` defining the OpenStack ID or names of properties for the CLOUD-state instances.
+
+Partitions/groups defining `cloud_nodes` may or may not also contain non-CLOUD state nodes (i.e. nodes in a matching inventory group). For CLOUD-state nodes, memory and CPU information is retrieved from OpenStack for the specified flavors. The `stackhpc.openhpc` group/partition options `ram_mb` and `ram_multiplier` and role variable `openhpc_ram_multiplier` are handled exactly as for non-CLOUD state nodes. This implies that if CLOUD and non-CLOUD state nodes are mixed in a single group all nodes must be homogenous in terms of processors/memory.
+
+Some examples are given below. Note that currently monitoring is not enabled for CLOUD-state nodes.
+
+## Example variable and partition definitions
+
+The example below could be placed in `environments/<env>/inventory/group_vars/openhpc/overrides.yml`:
+- Not shown here is the inventory group `dev_small` containing 2 (non-CLOUD state) nodes.
+- The "small" partition is the default and contains 2 non-CLOUD and 2 CLOUD nodes.
+- The "burst" partition contains only CLOUD-state nodes.
+
+```yaml
+openhpc_cluster_name: dev # normally in e.g. environments/<env>/inventory/hosts but shown here for clarity
+general_v1_small:
+  image: ohpc-compute-210909-1316.qcow2
+  flavor: general.v1.small
+  keypair: centos-at-steveb-ansible
+  network: stackhpc-ipv4-geneve
+
+general_v1_medium:
+  image: ohpc-compute-210909-1316.qcow2
+  flavor: general.v1.medium
+  keypair: centos-at-steveb-ansible
+  network: stackhpc-ipv4-geneve
+
+openhpc_slurm_partitions:
+- name: small
+  default: yes
+  cloud_nodes: dev-small-[2-3]
+  cloud_instances: "{{ general_v1_small }}"
+
+- name: burst
+  default: no
+  cloud_nodes: 'burst-[0-3]'
+  cloud_instances: "{{ general_v1_medium }}"
+```
+
+# Dependencies
+
+`stackhpc.openhpc` role. This role needs to run before tasks from `stackhpc.openhpc:runtime.yml` but requires the `slurm` user to be present. See example playbook below.
+
+# Example Playbook
+
+This role should be run on the Slurm controller only.
+
+```yaml
+  - hosts: control
+    become: yes
+    tasks:
+      - name: Install Slurm packages to create slurm user
```

Comment:
Interesting approach. I think this means we run install.yml twice. Initially, I thought that having a pattern where the roles could modify the slurm configuration in a separate set of tasks:
```
      - name: Inject openhpc configuration for auto scaling
        import_role:
          name: stackhpc.slurm_openstack_tools.autoscale
          tasks_from: openhpc_config
      - name: Setup Slurm
        import_role:
          name: stackhpc.openhpc
      - name: Setup autoscaling using OpenStack
        import_role:
          name: stackhpc.slurm_openstack_tools.autoscale
```
Might be more desirable as that would make it more obvious we were injecting config. But I guess slurm might check for the existence of the scripts configured in its config file, so I've come round to your way of doing things.

Reply:
That's exactly the problem - if you have a State=CLOUD node, and the resume/suspend scripts are not configured, not present, or have wrong permissions, slurmctld dies on startup. This is not great, but it is the best approach I could come up with.

We avoid running install.yml twice by changing the appliance like this:
https://github.com/stackhpc/ansible-slurm-appliance/blob/feature/autoscale/ansible/slurm.yml#L22